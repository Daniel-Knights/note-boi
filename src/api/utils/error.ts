import { AppError, ERROR_CODE, ErrorConfig } from '../../classes';

/** Catches hanging requests (e.g. due to server error). */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function throwFetchError<T extends (...args: any) => Promise<void>>(
  errorConfig: Omit<ErrorConfig<T>, 'message'>,
  originalError: unknown
): void {
  throw new AppError({
    ...errorConfig,
    message: 'Request failed',
    originalError,
  });
}

/** Catches note encryption errors. */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function throwEncryptorError<T extends (...args: any) => Promise<void>>(
  errorConfig: Omit<ErrorConfig<T>, 'code' | 'message'>,
  originalError: unknown
): void {
  throw new AppError({
    ...errorConfig,
    code: ERROR_CODE.ENCRYPTOR,
    message: 'Note encryption/decryption failed',
    originalError,
  });
}
