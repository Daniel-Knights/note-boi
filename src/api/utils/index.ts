export * from './decorators';
export * from './error';
export * from './http';
