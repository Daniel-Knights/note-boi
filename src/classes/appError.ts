export const ERROR_CODE = {
  NONE: 0,
  UNKNOWN: 1,
  LOGIN: 2,
  SIGNUP: 3,
  LOGOUT: 4,
  // PUSH was 5
  // PULL was 6
  DELETE_ACCOUNT: 7,
  CHANGE_PASSWORD: 8,
  ENCRYPTOR: 9,
  FORM_VALIDATION: 10,
  SYNC: 11,
} as const;

type ErrorCodes = typeof ERROR_CODE;
type ErrorKey = keyof ErrorCodes;
type ErrorCode = ErrorCodes[ErrorKey];

// Using `unknown` here breaks inference
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type ErrorConfig<T extends (...args: any) => void> = {
  code: ErrorCode;
  message?: string;
  originalError?: unknown;
  /** Function to call on 'Try again', and its arguments */
  retry?: {
    fn: T;
    args?: Parameters<T>;
  };
  /** Whether to display the error in a form and/or `SyncStatus` */
  display?: {
    form?: boolean;
    sync?: boolean;
  };
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class AppError<T extends (...args: any) => void = (...args: any) => void> {
  readonly code;
  readonly message;
  readonly display;
  readonly retryConfig;
  readonly originalError;

  constructor(config: ErrorConfig<T> = { code: ERROR_CODE.NONE }) {
    this.code = config.code;
    this.message = config.message;
    this.display = Object.freeze(config.display);
    this.retryConfig = Object.freeze(config.retry);
    this.originalError = config.originalError;

    this.retry = this.retry.bind(this);
  }

  get isNone() {
    return this.code === ERROR_CODE.NONE;
  }

  retry() {
    if (!this.retryConfig) return;

    const { fn, args } = this.retryConfig;

    return args ? fn(args) : fn();
  }
}
