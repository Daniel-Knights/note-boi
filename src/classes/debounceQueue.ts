/**
 * Queues and debounces async function calls.
 * If a function is cancelled after it has already begun executing, `isCancelled` can
 * be used to check if it should stop.
 */
export class DebounceQueue {
  #cancelled = new Map<number, { isRunning: boolean }>();
  #current: { id: number | null; isRunning: boolean } = {
    id: null,
    isRunning: false,
  };

  /**
   * Adds new task to the debounce queue.
   * If there is an existing task, it will be cleared.
   */
  add(cb: (isCancelled: () => boolean) => Promise<void>, delay?: number) {
    this.clear();

    const timeoutId = window.setTimeout(() => {
      if (timeoutId !== this.#current.id) return;

      this.#current.isRunning = true;

      cb(() => this.#isCancelled(timeoutId)).finally(() => {
        this.#cancelled.set(timeoutId, { isRunning: false });
      });
    }, delay);

    this.#current.id = timeoutId;
    this.#current.isRunning = false;
  }

  /**
   * Clears the current task in the debounce queue.
   * If the task is running, it will be marked as cancelled.
   */
  clear() {
    if (!this.#current.id) return;

    if (this.#current.isRunning) {
      this.#cancelled.set(this.#current.id, { isRunning: true });
    } else {
      window.clearTimeout(this.#current.id);

      this.#current.id = null;
    }
  }

  /**
   * Checks if task with the given timeout ID has been cancelled.
   */
  #isCancelled(timeoutId: number) {
    const cancelled = !!this.#cancelled.get(timeoutId);

    this.#cancelled.delete(timeoutId);

    return cancelled;
  }
}
