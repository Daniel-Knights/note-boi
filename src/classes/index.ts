export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './note';
export * from './storage';
export * from './unsyncedNotesManager';
