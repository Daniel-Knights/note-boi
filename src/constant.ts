export const NOTE_EVENTS = {
  new: 'note-new',
  /** For preventing note edit on text change. */
  select: 'note-select',
  /** For whenever the content of a note is changed. */
  change: 'note-change',
  unsynced: 'note-unsynced',
};

export const STORAGE_KEYS = {
  USERNAME: 'username',
  TOKEN: 'token',
  UNSYNCED: 'unsynced-note-ids',
  THEME: 'theme',
  MENU_WIDTH: 'note-menu-width',
};
