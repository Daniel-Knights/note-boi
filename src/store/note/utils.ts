import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
