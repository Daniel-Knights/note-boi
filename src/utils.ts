import { event, invoke } from '@tauri-apps/api';
import { EventCallback, UnlistenFn } from '@tauri-apps/api/event';

import { TauriCommand } from './constant';
import type { Note } from './store/note';

/** `process.env.NODE_ENV === 'development'`. */
export function isDev(): boolean {
  return process.env.NODE_ENV === 'development';
}

/** Formats Unix time to date-time. */
export function unixToDateTime(unixTime: number): string {
  return Intl.DateTimeFormat([], {
    dateStyle: 'medium',
    timeStyle: isDev() ? 'long' : 'short',
  }).format(unixTime);
}

/** Gets key value from `localStorage` and parses it as JSON. */
export function localStorageParse(key: string): ReturnType<typeof JSON.parse> {
  return JSON.parse(localStorage.getItem(key) || '{}');
}

/** Returns `true` if string consists of only whitespace characters or is empty. */
export function isWhitespaceOnly(text?: string): boolean {
  return text?.trim() === '';
}

/** Returns true if note content is empty or note is `undefined`. */
export function isEmptyNote(note?: Note): boolean {
  if (!note) return true;

  return isWhitespaceOnly(note.content.title) && isWhitespaceOnly(note.content.body);
}

/** Calls {@link event.emit}, with stronger typing for `id`. */
export function tauriEmit<T>(id: 'login' | 'logout', payload?: T): void {
  event.emit(id, payload);
}

/** Calls {@link event.listen}, with stronger typing for `id`. */
export function tauriListen(
  id:
    | 'reload'
    | 'new-note'
    | 'delete-note'
    | 'push-notes'
    | 'login'
    | 'logout'
    | 'signup'
    | `tauri://${string}`,
  cb: EventCallback<unknown>
): Promise<UnlistenFn> {
  return event.listen(id, cb);
}

/** Calls {@link invoke}, with stronger typing for `cmd`. */
export function tauriInvoke<T>(
  cmd: TauriCommand,
  args?: Parameters<typeof invoke>[1]
): Promise<T> {
  return invoke(cmd, args);
}
