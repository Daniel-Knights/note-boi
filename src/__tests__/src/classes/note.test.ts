import { Note } from '../../../classes';
import { UUID_REGEX } from '../../constant';
import { floorToThousand } from '../../utils';

it('new Note()', () => {
  const emptyNote = new Note();
  const timestamp = Date.now();

  assert.strictEqual(typeof emptyNote.uuid, 'string');
  assert.lengthOf(emptyNote.uuid, 36);
  assert.isTrue(UUID_REGEX.test(emptyNote.uuid));
  assert.strictEqual(floorToThousand(emptyNote.timestamp), floorToThousand(timestamp));
  assert.deepEqual(emptyNote.content.delta, {
    ops: [],
  });
  assert.strictEqual(emptyNote.content.title, '');
  assert.strictEqual(emptyNote.content.body, '');
});
