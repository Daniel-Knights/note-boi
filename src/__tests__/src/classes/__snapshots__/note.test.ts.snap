// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`new Note() > KeyStore.getKey() 1`] = `undefined`;

exports[`new Note() > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`new Note() > Storage.get('THEME') 1`] = `null`;

exports[`new Note() > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`new Note() > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`new Note() > Storage.get('USERNAME') 1`] = `null`;

exports[`new Note() > allCalls 1`] = `[]`;

exports[`new Note() > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`new Note() > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`new Note() > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`new Note() > openedPopup 1`] = `undefined`;

exports[`new Note() > selectedTheme 1`] = `"System"`;

exports[`new Note() > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`new Note() > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
